fastly/block_fastly_service_v1_logging_openstack.go

```diff
@@ -0,0 +1,289 @@
+package fastly
+
+import (
+	"fmt"
+	"log"
+
+	gofastly "github.com/fastly/go-fastly/fastly"
+	"github.com/hashicorp/terraform-plugin-sdk/helper/schema"
+)
+
+type OpenstackServiceAttributeHandler struct {
+	*DefaultServiceAttributeHandler
+}
+
+func NewServiceLoggingOpenstack() ServiceAttributeDefinition {
+	return &OpenstackServiceAttributeHandler{
+		&DefaultServiceAttributeHandler{
+			key: "logging_openstack",
+		},
+	}
+}
+
+func (h *OpenstackServiceAttributeHandler) Process(d *schema.ResourceData, latestVersion int, conn *gofastly.Client) error {
+	serviceID := d.Id()
+	ol, nl := d.GetChange(h.GetKey())
+
+	if ol == nil {
+		ol = new(schema.Set)
+	}
+	if nl == nil {
+		nl = new(schema.Set)
+	}
+
+	ols := ol.(*schema.Set)
+	nls := nl.(*schema.Set)
+
+	removeOpenstackLogging := ols.Difference(nls).List()
+	addOpenstackLogging := nls.Difference(ols).List()
+
+	// DELETE old OpenStack logging endpoints.
+	for _, oRaw := range removeOpenstackLogging {
+		of := oRaw.(map[string]interface{})
+		opts := buildDeleteOpenstack(of, serviceID, latestVersion)
+
+		log.Printf("[DEBUG] Fastly OpenStack logging endpoint removal opts: %#v", opts)
+
+		if err := deleteOpenstack(conn, opts); err != nil {
+			return err
+		}
+	}
+
+	// POST new/updated OpenStack logging endpoints.
+	for _, nRaw := range addOpenstackLogging {
+		lf := nRaw.(map[string]interface{})
```

Comment:
To use the custom `StateFunc` and to ensure that it works when there is a trailing new line, add the following:

```suggestion
		lf := nRaw.(map[string]interface{})
		
		// @HACK for a TF SDK Issue.
		//
		// This ensures that the required, `name`, field is present.
		//
		// If we have made it this far and `name` is not present, it is most-likely due
		// to a defunct diff as noted here - https://github.com/hashicorp/terraform-plugin-sdk/issues/160#issuecomment-522935697.
		//
		// This is caused by using a StateFunc in a nested TypeSet. While the StateFunc
		// properly handles setting state with the StateFunc, it returns extra entries
		// during state Gets, specifically `GetChange("logging_kafka")` in this case.
		if v, ok := cfg["name"]; !ok || v.(string) == "" {
			continue
		}
```